### PR TITLE
🤖 feat: share the workspace memory notebook across a sub-agent task tree

### PR DESCRIPTION
**Pull request by ThomasK33**
## Summary

Sub-agent workspaces (those spawned with a `parentWorkspaceId`) now read and write the memory notebook of the root workspace of their task tree instead of a private one. `/memories/workspace` for a child resolves to `<sessions>/<owner>/memory`, the sidecar and refinement journal are keyed by the owner, the Memory tab and open sessions follow ownership changes, and Dream (consolidation) runs for a child are redirected to the owner. Transcripts and other session artifacts stay per workspace.

Stack: layer 1 of the split of #4139. This layer is the "sharing works while every workspace is alive" slice. Removal and rollback safety, memory-policy epochs, legacy note adoption, and multi-backend coherence follow as stacked PRs.

## Background

Sub-agents forked from a workspace could not see the parent's notebook, and their own notes were lost when the child was removed. #4139 implemented sharing but grew to +15.8k lines over 91 review rounds by fixing every finding in place. It is being re-delivered as a stack carved from that final tree, so each layer is reviewable on its own.

## Implementation

1. `memoryWorkspaceOwner.ts`: `resolveWorkspaceMemoryOwnerId` walks the `parentWorkspaceId` chain over one config snapshot (indexed once per snapshot via a WeakMap). Unknown IDs, dangling parents, cycles, and depth overflow resolve to the workspace itself so a misconfigured tree degrades to per-workspace behavior.
2. `MemoryService`: owner memo invalidated by a config file stamp (`Config.configFileStamp`) and by `onConfigChanged`; the owner resolved at the start of a command is pinned per command context, and `assertMutationCommittable` refuses to commit if ownership changed mid-command (for example `config.json` recovered from unreadable and the workspace moved from its private fallback to the shared store). Store paths, sidecar logical keys, change events, and refinement journal rows are owner keyed. `setPinned` runs under the store mutation lock; Dream's delete/rename pass `rejectPinned` so the pin check happens inside the lock.
3. Subscriptions and sessions: `subscribeMemoryChanges` matches workspace-scope events by resolved owner and re-roots on `ownersInvalidated`; `di/layers/core.ts` invalidates the cached memory context of every session in the tree on a write or ownership change.
4. Consolidation: `maybeRun` redirects a child to its owner (teardown and archive gates checked first), `runCore` and the launch sweep skip children, harvest recovery iterates all tree members, and status reports the owner's records.

## Known limitations (addressed in later layers)

1. Removing a workspace whose store is shared is not yet coordinated with the owner lock, and grandchildren of a removed intermediate ancestor fall back to their private store (L1b).
2. Notes a sub-agent wrote before this change stay in its private store and are not read (legacy adoption, L3).
3. A second backend on the same `~/.xum` is not detected (L4).

## Validation

Ported tests cover owner resolution edge cases, store redirection, journal placement, config-change re-resolution (local edit, external rewrite, unreadable then recovered), refusal to commit after an ownership change, subscription owner matching and invalidation, and Dream redirection/skip. Sibling suites (memory, consolidation, sessions, workspace service/removal, task service, refinement rollback, config): 1698 pass.

## Risks

Medium, scoped to workspaces with a `parentWorkspaceId`. Workspaces without one resolve to themselves and keep today's behavior. The main regression surface is the owner memo invalidation path (a stale owner would write to the wrong store); it is guarded by the in-lock ownership re-check.

---

_Generated with `xum` • Model: `anthropic:claude-fable-5-1` • Thinking: `high` • Cost: `$956.36`_

### CODE REVIEW

**Comment by ThomasK33** [comment]
@codex review

File: src/node/services/memoryService.ts [hunk 1]
```diff
@@ -804,6 +994,60 @@ export class MemoryService extends EventEmitter {
...
+      if (currentOwner !== boundOwner) {
+        throw new MemoryCommandError(
+          `Ownership of the workspace notebook changed while mutating ${virtualPath} (now ${currentOwner}); retry the command`
+        );
+      }
+    }
+    if (await isWorkspaceRemovalTombstoned(this.config.rootDir, ctx.workspaceId)) {
+      throw new MemoryCommandError(
+        `Workspace ${ctx.workspaceId} was removed; refusing to commit the mutation of ${virtualPath}`
+      );
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/memoryService.ts]: Fixed in 4d61dceb2: assertMutationCommittable now checks the store owner's removal tombstone as well as the actor's for workspace-scope mutations (global/project stores are unaffected). Test: "refuses a sub-agent's mutation once the owner's removal tombstone exists".

File: src/node/services/di/layers/core.ts [hunk 2]
```diff
@@ -573,6 +573,25 @@ export const CoreWiringLive: Layer.Layer<
...
+    // every live session resolving to that owner reads the same notebook.
+    memoryService.on("change", (event: MemoryChangeEvent) => {
+      if (event.scope !== "workspace" || event.workspaceId === "") return;
+      // One config snapshot for the whole pass: cold owner lookups would
+      // otherwise parse the config once per live session, synchronously.
+      let cfg: ReturnType<typeof config.loadConfigOrDefault> | undefined;
+      const snapshot = () => (cfg ??= config.loadConfigOrDefault());
+      workspaceService.invalidateMemoryContextWhere(
+        (workspaceId) =>
+          memoryService.resolveWorkspaceMemoryOwnerId(workspaceId, snapshot) === event.workspaceId
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/di/layers/core.ts:586]: Fixed in 4d61dceb2: recordUsage now emits the owner-addressed change event for workspace-scope reads (view, recall), so the core.ts listener invalidates every tree session's cached memory context. The store-revision half of the final design stays in the multi-backend layer (L4).

File: src/node/services/memoryConsolidationService.ts [hunk 3]
```diff
@@ -698,7 +774,10 @@ export class MemoryConsolidationService extends EventEmitter {
       this.inFlight.delete(workspaceId);
       removal.dispose();
     }
-    if (options.skipHarvestRecovery !== true) {
+    // Removal cancelled this run: recovery would spawn child harvests outside
+    // the cancellation registry being drained (their controllers were never
+    // registered), mutating the shared inbox during destructive teardown.
+    if (options.skipHarvestRecovery !== true && !this.removalCancelled.has(workspaceId)) {
       await Effect.runPromise(this.recoverRetryableHarvestsEffect(workspaceId));
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/memoryConsolidationService.ts]: Fixed in 4d61dceb2: harvest recovery after a run is gated on options.actingWorkspaceId being cancelled as well as the owner. Test extended with a sibling's retryable record that must stay at attemptCount 1 after the acting child's cancellation.

File: src/node/services/memoryService.ts [hunk 4]
```diff
@@ -717,15 +899,22 @@ export class MemoryService extends EventEmitter {
           );
         }
         return new LocalMemoryStore(
-          workspaceMemoryStorePath(this.config.sessionsDir, ctx.workspaceId)
+          workspaceMemoryStorePath(this.config.sessionsDir, this.ownerWorkspaceIdFor(ctx))
         );
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/memoryService.ts:910]: Deferred to the legacy-adoption layer of this stack (L3), as listed under Known limitations in the PR body. The adoption pass (manifest, staging directory, receipts, conflict targets under imported/<child>/) is a new module plus on-disk artifacts, so per the review-scope rule it lands as its own stacked PR rather than growing this one. Until then pre-upgrade child notes stay in place on disk and are not read; nothing is deleted.

File: src/node/services/memoryWorkspaceOwner.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      if (entry === undefined) {
+        // Only the chain root may be unknown without invalidating the walk: an
+        // unregistered starting workspace simply resolves to itself.
+        if (current !== workspaceId) {
+          log.warn("[memory] parentWorkspaceId points at an unknown workspace", {
+            workspaceId,
+            parentWorkspaceId: current,
+          });
+        }
+        return workspaceId;
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/memoryWorkspaceOwner.ts:69]: Deferred to the next stacked PR (L1b, removal and rollback safety): the fix pins surviving descendants to their current owner at removal time, which needs the persisted memoryOwnerWorkspaceId workspace field. In this layer a grandchild of a removed intermediate workspace falls back to its private store (fail-safe, no cross-tree redirect); listed under Known limitations.

File: src/node/services/memoryService.ts [hunk 6]
```diff
@@ -769,7 +958,8 @@ export class MemoryService extends EventEmitter {
    * Append the invertible `refinement` row for one memory mutation (RLM r2).
    *
    * Rows land in the ACTING workspace's session journal even though memory
-   * files can be global/project-scoped: the journal is per-session, so
+   * files can be global/project-scoped — or, for a sub-agent's workspace
+   * scope, live in the OWNER's session dir: the journal is per-session, so
    * cross-workspace edits to a shared file are attributed to (and invertible
    * from) whichever workspace made them — the intended v1 scope. When the
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/memoryService.ts:971]: Deferred to the next stacked PR (L1b): rollbackRefinement gains the securely resolved owner root (sharedWorkspaceMemorySessionDir plus the tree's peer journals) in the rollback engine, the refinement_rollback tool, and the debug CLI. This layer keeps the child-journal attribution and fails closed: the rollback is refused rather than misapplied, which is what the test pins.

**Comment by ThomasK33** [comment]
@codex review

File: src/node/services/memoryService.ts [hunk 7]
```diff
@@ -717,15 +906,22 @@ export class MemoryService extends EventEmitter {
           );
         }
         return new LocalMemoryStore(
-          workspaceMemoryStorePath(this.config.sessionsDir, ctx.workspaceId)
+          workspaceMemoryStorePath(this.config.sessionsDir, this.ownerWorkspaceIdFor(ctx))
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/memoryService.ts:909]: Fixed in 6fa91ef21: removeSessionDirUnderMemoryLocks takes an optional sharedWorkspaceMemorySessionDir and adds the owner store's mutation lock key to the removal lock set; removeWorkspace resolves the child's owner (still registered at that point) and passes it. Test: "sub-agent removal also waits for a writer holding the OWNER's memory store lock".

File: src/node/services/agentSession.ts [hunk 8]
```diff
@@ -1049,6 +1049,17 @@ export class AgentSession {
...
   private memoryContextByModelString = new Map<string, CachedMemoryContext>();
+
+  /**
+   * Drop the cached memory context so the next stream rebuilds the index and
+   * hot set from disk. Own memory tool calls clear it on tool-call-end; this
+   * entry point is for writes by OTHER sessions to a store this session also
+   * reads (a sub-agent editing the task tree's shared workspace notes).
+   */
+  invalidateMemoryContext(): void {
+    this.memoryContextByModelString.clear();
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/agentSession.ts:1060]: Fixed in 6fa91ef21: invalidateMemoryContext bumps a generation counter; resolveMemoryContext skips caching a snapshot whose build overlapped an invalidation, and prepareRolloverRequest's start() installs the staged map only if the generation is unchanged (otherwise a fresh map). Test: "does not cache a context whose build overlapped an invalidation".

**Comment by ThomasK33** [comment]
@codex review

File: src/node/services/memoryService.ts [hunk 9]
```diff
@@ -627,6 +809,13 @@ export class MemoryService extends EventEmitter {
       const key = this.logicalKeyFor(ctx, scope, relPath);
       if (key === null) return;
       await this.metaService.recordAccess(key, options);
+      if (scope === "workspace" && !options.write) {
+        // A read-side access (view, recall) re-ranks the shared hot set the
+        // whole task tree derives from the owner's sidecar entries, so it is
+        // published like a pin: the other live sessions of the tree drop
+        // their cached memory context. Writes publish with their mutation.
+        this.emitChange(ctx, scope, relPath, "agent");
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/memoryService.ts:817]: Fixed in the stack top layer, #4250 (commits b2ca6a2bc..ecaf896f6): read-side re-ranking events now carry reason: access on the memory change event and its subscription schema; MemoryBrowser still refreshes on them but no longer adds the path to agentEditedPaths. Mutation events are unchanged.

File: src/node/services/memoryService.ts [hunk 10]
```diff
@@ -608,15 +633,172 @@ export class MemoryService extends EventEmitter {
...
+    workspaceId: string,
+    snapshot?: () => ReturnType<Config["loadConfigOrDefault"]>
+  ): string {
+    if (snapshot !== undefined) return resolveWorkspaceMemoryOwnerId(snapshot(), workspaceId);
+    const stamp = this.config.configFileStamp();
+    if (stamp !== this.workspaceMemoryOwnerConfigStamp) {
+      // Adopt the new stamp only once its contents were actually read: a
+      // changed-but-unreadable file (see below) must be retried on the next
+      // call, not remembered as "seen".
+      if (this.invalidateWorkspaceMemoryOwnerMemo()) this.workspaceMemoryOwnerConfigStamp = stamp;
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/node/services/memoryService.ts:669]: Deferred to the multi-backend layer (L4) of this stack split, together with the throttled fingerprint and sidecar-locking items already deferred there. Closing it needs a cross-process ownership signal or a durable ownership stamp check on the AgentSession context cache, which is a new mechanism beyond this layer; the single-backend path is covered by the stamp check in resolveWorkspaceMemoryOwnerId and the ownersInvalidated event.
